Today is that day - Single pass through Calcite planner

paul-rogers opened the pull request:
The Druid planner contains the following comment:

```text
   * In some future this could perhaps re-use some of the work done by {@link #validate(boolean)}
   * instead of repeating it, but that day is not today.
```

Well, today *is* that day. The Druid planner now makes only one pass through Calcite planner. This means we parse and validate the query only once, not twice as before.

The original two-pass approach likely was a way to resolve a conflict. Druid wants access to the "validator" used to validate a query so we can extract the "resource actions" (the datasources which the query accesses.) The validator is private to the default `CalcitePlanner`, so Druid created its own. However, to then continue on to planning, the Calcite planner throws a fit because it thinks we skipped the validation step, because we didn't do validation though Calcites's own default planner. Hence, we had to start over and do it Calcite's way.

Now, as it turns out, there is nothing special about the default `CalcitePlannerImpl`: it seems to exist as a handy way to support out-of-the-box, JDBC-based queries. More advanced use cases generally provide their own implementation. So, this PR clones the Calcite class, but with adjustments for Druid. The main adjustment is to provide access to the validator.

Once we resolve the validation issue, the rest is just plumbing: saving the parser state between validate on the one hand, and prepare/plan on the other. `SqlLifecycle` and `DruidPlanner` change a bit to handle that task.

The PR also moves some validation-like code from the `plan()` method to `validate()` to better fit into the new flow.

The key risk with this kind of change is that we break something. To catch any regression, this work was done in a private branch that also had the [planner test framework](https://github.com/apache/druid/pull/12545). The planner artifacts (schema, logical plan, native query) were identical before and after the change. The various `Calcite?QueryTest` cases provide a lighter validation in this PR itself, since the planner framework is not yet in master, nor is it included in this PR.

The result of this change is that:

* The Druid planner produces less "garbage" due to Calcite's heavy object creation.
* We do half the work within Calcite, thus saving some CPU resource.
* Table (datasource) checks are done once, avoiding potential race conditions if they change between validation and plan.

### Authorization

Review comments called attention to some oddities in authorization that are addressed in a revised commit:

* Authorization is moved from `SqlLifecycle` to `DruidPlanner` so it can be enforced as part of the query lifecycle. A new `authorize()` call does the job and ensures that the call is a) made, and b) made at the correct time.
* Because of this, `ValidationResult` became redundant and was removed: the resource actions are mostly now private to the planner itself.

<hr>

This PR has:
- [X] been self-reviewed.
- [X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] been tested in a test Druid cluster.

clintropolis commented:
heh, I think I wrote that comment 😅 

iirc the main reason I didn't make the change at the time is because I didn't want to validate that stateful jdbc connections actually worked if we made them be _actually_ stateful (instead of repeating the work per request). So this is probably the area where most testing should focus, especially on prepared statements with dynamic parameters.

IIRC, for dynamic parameters, I think we do actually want to re-run through the planner once we have values for the actual parameter bindings because some stuff can be optimized during planning, but... it was a while ago when I wired them up so maybe its not an issue?

Anyway, I'll try to have a closer look at this sometime soon :+1:

paul-rogers commented:
@gianm, thanks for the background. The new planner test framework will catch the kinds of issues you mention: it captures the details of the Calcite logical plan (as well as the native query, like the existing tests do.) The good news is that the original version of these changes were done in a branch that had the new planner tests, and they reported no changes in the Calcite artifacts.

That said, it is worth spending time to trace exactly how we handle parameters to be sure we account for the proper flow.

In general, there should be no reason to replan a query under normal conditions: other tools that use Calcite generally make do with a single pass. Where tools tend to plan a second time is if the planner paints itself into a corner, some global state is changed, and the planner is run a different way. (Impala used to do that.)

clintropolis commented:
>That said, it is worth spending time to trace exactly how we handle parameters to be sure we account for the proper flow.

Dynamic parameters are handled in two places, the first by rewriting the `SqlNode` with https://github.com/apache/druid/blob/master/sql/src/main/java/org/apache/druid/sql/calcite/planner/SqlParameterizerShuttle.java#L40 and the second, if any of them are missed at that layer, at the `RelNode` level with https://github.com/apache/druid/blob/master/sql/src/main/java/org/apache/druid/sql/calcite/planner/RelParameterizerShuttle.java#L59. The PR that added them, #6974, has some additional details.

> In general, there should be no reason to replan a query under normal conditions: other tools that use Calcite generally make do with a single pass. Where tools tend to plan a second time is if the planner paints itself into a corner, some global state is changed, and the planner is run a different way. (Impala used to do that.)

I somewhat remember getting different results based on whether or not the parameter bindings (and more importantly, their types) were present or not when running things through the planner, but if I remember might not change the end result native query so much. But yeah, in the JDBC case the state is a bit different once the parameters have values bound to them because they now have a type. SQL over HTTP also supports dynamic parameters, but in that case the bindings come with the HTTP request, instead of after preparing the statement like in JDBC, so there is no intermediary state where we have a bunch of typeless things hanging out.

paul-rogers commented:
@clintropolis, thanks again for the description; it makes sense. @clintropolis described offline the process we use for parameters. Basically, when a query is to be run, and parameters provides along with the query text, `SqlParameterizerShuttle` substitutes in values early in the planing process for the reason @clintropolis outlines in the items linked above. `SqlParameterizerShuttle` may skip some parameters (such as `DATE` types with integer values) that it cannot handle.

If the query comes via Avatica, and is a PREPARE, then there are no parameters values: parameters are left as placeholders with values to be filled in per-query.

In either case, in the `plan()` step, any remaining parameters are replaced by values in `RelParameterizerShuttle`. This handles the case either of Avatica running a prepared query, or those pesky values which `SqlParameterizerShuttle` could not convert.

Note also that in the query-with-parameters case, `SqlParameterizerShuttle` checked that values were provided. In the Avatica PREPARE case, there were no values, and `SqlParameterizerShuttle` wasn't called. In this case, it falls to `RelParameterizerShuttle` to do the check that the required values exist.

As @clintropolis explained, the `RelParameterizerShuttle` check is only done for the "Druid convention" path. The "bindable convention" has its own parameter binding and checks.

Given all this (and after a bit more code tweaking and comment-adding), the one-pass approach in this PR seems to handle all of the above paths through the planner.

gianm commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/CalcitePlanner.java:
```diff
@@ -0,0 +1,432 @@
+import java.util.Properties;
+
+/**
+ * Calcite planner. Clone of Calcite's {@code SqlPlannerImpl},
```
Clone of Calcite's SqlPlannerImpl as of what version? Including that here will make it easier for people in the future to see if there's any useful changes from SqlPlannerImpl that need to be merged back into our version.

Also I think it's called PlannerImpl.

gianm commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/CalcitePlanner.java:
```diff
@@ -0,0 +1,432 @@
+ * Calcite planner. Clone of Calcite's {@code SqlPlannerImpl},
+ * but with the validator made accessible.
+ */
+public class CalcitePlanner implements Planner, ViewExpander
```
I'm surprised checkstyle accepted this file. We must have pretty loose rules, since it doesn't look to me like a normal Druid source file. I suppose that's OK.

gianm commented on sql/src/main/java/org/apache/druid/sql/calcite/view/DruidViewMacro.java:
```diff
@@ -58,6 +58,7 @@ public TranslatableTable apply(final List<Object> arguments)
   {
     final RelDataType rowType;
     try (final DruidPlanner planner = plannerFactory.createPlanner(viewSql, new QueryContext())) {
+      planner.validate(false);
```
I didn't realize that view expansion was done with an empty query context. I suppose this makes sense: it ensures the view is the same for everyone.

IMO, it's better to set this to always `true` instead of always `false`. Right now, it doesn't matter either way, because the context is empty. But if it's ever made nonempty, then `true` is a safer constant value because it defaults to the more restrictive security mode. If that doesn't make sense in the future, then the future person that is adding the context parameters can sort out what's best.

A comment would be helpful too, like:

```
// Since the context is empty, it doesn't matter if the authorizeContextParams flag is true or false.
```

clintropolis replied:
>As part of the auth cleanup, I moved the authorizeContextParams flag to the authorization step, and clearly stated that a statement (or view) can be prepared without authorization. Only execution (AKA "plan") needs authorization.

Hmm, currently `DruidStatement` calls `validateAndAuthorize` for both `prepare` and `execute`, are you saying that `prepare` no longer authorizes or that it never did? (because looking at the current code it seems to do 

>The standard way to handle views is to assign them an owner. Views run with the owner's permissions. Queries that use the view must be authorized against the view, not against the resources which the view uses. Of course, Druid has no concept of users, so the idea of "owner" is ill-defined. Maybe we stash a bundle of permissions with the view or some such? That's a question for another time.

There is a 'VIEW' resource that views are authorized against, not the resource that the view query uses, which i think is more or less the same thing?

paul-rogers replied:
What I saw in the code is that the Avatica path did validate, authorize, prepare. But, the view path did validate, prepare but no authorize (it has no auth result or request). The flow before and after the change is the same. For Avatica, the  path is validate, authorize, prepare. The query path is validate, authorized, plan. The view path is validate, prepare (with no authorize). Does this sound right?

clintropolis replied:
Oh, I misunderstood what you were saying. Yes within the view's usage of the planner it is basically unauthorized on purpose, because the view itself as a whole is authorized as a `ResourceType.VIEW` by the `SqlLifecycle` of the query using the view. The planner usage here is to get the row type information so the query planner using the view can do its thing. #10812 has details on view authorization stuff (which is also the PR that left the comments that this PR is all about).

gianm commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/SqlParameterizerShuttle.java:
```diff
@@ -50,34 +63,36 @@ public SqlParameterizerShuttle(PlannerContext plannerContext)
+        );
+      }
+
+      // This throws ClassCastException for a DATE parameter given as
```
If this is the only place that's expected to throw a ClassCastException, how about wrapping only this particular line in the try/catch?

gianm commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:
```diff
@@ -333,6 +371,9 @@ private PlannerResult planWithDruidConvention(
                           .filter(action -> action.getAction() == Action.READ)
                           .collect(Collectors.toSet());
 
+        // TODO: This is not really a state check since there is a race condition.
```
What's the TODO here — I don't see what change needs to be made later?

And, whatever that change is, why not make it now?

paul-rogers replied:
That rework suggested that there are several other areas to clean up, and that `SqlLifecycleTest` should be rewritten. That will come later, to limit the blast radius of this PR.

gianm reviewed (commented):
Very nice to have this cleaned up! The changes mostly look good to me. I had a few line comments. I didn't review the code that was marked as copied from Calcite, because I assume that it's the same stuff we're already using.

I also am not familiar with the JDBC / dynamic parameter issue that you and @clintropolis have been discussing, so I didn't consider that particular situation in my review. Do the tests cover it well enough?

paul-rogers commented:
@clintropolis, thanks for the history of parameters. I reread the links that you generously provided: very helpful. @gianm your advice to look into the JDBC (Avatica) code path was wise: doing so was insightful.

It turns out that Druid's implementation of the JDBC protocol is a bit off, which may have led to some of the confusion around when to prepare and when to expect parameters. Issue #12682 spells out the gaps.

This latest PR revises how we manage SQL processing in both HTTP and JDBC to try to get things a bit cleaner. A key challenge is that `SqlLifecycle` tries to implement two distinct processing flows, while strictly enforcing a set of states. This has caused the code to get rather complex. According to the comments, `SqlLifecycle` evolved from the native `QueryLifecycle` optimize for native queries. But SQL, for better or worse, works differently.

SQL works around the idea of _statements_ and _cursors_. In JDBC, statements are of two kinds: `Statement` and `PreparedStatement`. The cursor is represented by a `ResultSet`. Each statement can have one or more result sets open, and each statement can give rise to any number of result sets (one after another.) In a prepared statement, each execution tends to take a set of parameter values.

Then, on the HTTP side, we have a "direct execution" model: we get a SQL statement, context variables and parameters, and run through the planner sequence: parse, validate, authorize, plan and execute.

To model this behavior, the latest commit refactors `SqlLifecycle` in to focused "statement" classes:

* `DirectStatement`: Implements a one-pass, directly executed statement. Accepts parameters. Makes a single pass through the planner. Produces a `Sequence` (Druid's result set) as output.
* `HttpStatement`: A direct statement for use in the HTTP endpoint.
* `PreparedStatement`: The Druid implementation of the JDBC prepared statement. Produces a `PrepareResult` as output. Allows execution by creating a `DirectStatement` which runs the query, with parameters. Supports the JDBC prepared statement model.

The resulting code is far simpler and easier to test. Check out the result to see if the refactoring makes sense, and resolves the issues we've been discussing.

Then, the Avatica JDBC implementation is refactored to clearly model the JDBC concepts:

* `DruidJdbcStatement` models a (non-prepared) `Statement`.
* `DruidJdbcPreparedStatement` models a `PreparedStatement`.
* `AbstractDruidJdbcStatement` is the common base class for the above.
* `DruidJdbcResultSet` represents a result set (a running query).

The JDBC classes are then implemented using the statement abstractions described above. A JDBC `ResultSet` is a `DirectStatement`. A JDBC `PreparedStatement` is a Druid `PreparedStatement`.

The Avatica driver tests were modified to be consistent with the JDBC standard. The original code didn't pass, but the revised code does.

The result is that we now have a clean SQL stack for JDBC, HTTP and internal use that make a single planner pass when possible, and use two passes for the JDBC prepare/execute model.

Please take a look and see what you think.

paul-rogers commented:
Also fixes a minor issue with context variable security: it turns out we were using context variables to determine if someone can cancel a query. See issue #12683. Fixed so that only table permissions (datasources, views, etc.) control whether someone can cancel a query as we would expect.

paul-rogers commented:
Grew to large. WIll split out supporting work into separate PRs.

paul-rogers commented:
Rebased on the latest master which incorporates some of the changes previously here. Squashed commits to make the merge easier. There are no new changes in this latest commit: only removals of those files committed elsewhere (and removing the later `SqlLifecycle` refactoring to put the code back to the state it was at the prior reviews.)

clintropolis commented on sql/src/main/java/org/apache/druid/sql/SqlLifecycle.java:
```diff
@@ -248,46 +256,44 @@ private ValidationResult validate(AuthenticationResult authenticationResult)
   {
+    Access authorizationResult = planner.authorize(
+        authorizer,
+        authConfig.authorizeQueryContextParams());
```
super nit: formatting, the last `)` should be on the following line, this is one of those edge cases the style bots don't catch for whatever reason (i seem to recall it being some limitation in .. something.. somewhere).

clintropolis commented on sql/src/main/java/org/apache/druid/sql/calcite/view/DruidViewMacro.java:
```diff
@@ -58,7 +58,8 @@ public TranslatableTable apply(final List<Object> arguments)
     try (final DruidPlanner planner = plannerFactory.createPlanner(viewSql, new QueryContext())) {
-      rowType = planner.plan().rowType();
+      planner.validate();
+      rowType = planner.prepare().getRowType();
```
any reason to switch to prepare row type or just to do a bit less work?

paul-rogers replied:
I *think* this is due to the state machine added to the planner. It enforces that authorization is done before planning. However, in this context, we have no auth context. So, by using prepare instead, we get the information without the need to authorize.

And, as you point out, the job of prepare is to provide this info: planning (to get a physical execution plan) is overkill.

clintropolis commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java:
```diff
@@ -0,0 +1,65 @@
+ * Clone of Calcite's {@code CalciteSqlValidator} which is not
+ * visible to Druid.
+ */
+class DruidSqlValidator extends SqlValidatorImpl
```
hmm, should this be instead modifications to https://github.com/apache/druid/blob/master/sql/src/main/java/org/apache/calcite/prepare/DruidSqlValidator.java or does the `DruidPlanner` really need to clone the validator differently than the other one used by `CalcitePlanner`?

paul-rogers replied:
Good catch. So, what's *supposed* to happen is that the now-renamed `BaseDruidSqlValidator` simply exposes the Java-specific `CalciteSqlValidator` by living in the Calcite name space and extending a protected class. Then, `DruidSqlValidator` is the Druid version, with soon-to-be-added extensions for the `INSERT` statement. The latest commit makes this clearer: `DruidSqlValidator` extends `BaseDruidSqlValidator` and removes the methods we now inherit from `CalciteSqlValidator`. Clear as mud?

Let's let Travis do its thing to validate this latest adjustment, then please take another look.

clintropolis reviewed (commented):
lgtm overall

paul-rogers commented:
Argh... Tests that use mocks to enforce a call sequence of a class's implementation are quite tedious, and don't shed much light. Adjusted `SqlLifecycleTest` for the revised code flow. Luckily, this test won't be with us much longer; a later PR in this sequence will escort it off to the great Bit Bucket in the sky.

clintropolis commented:
>Argh... Tests that use mocks to enforce a call sequence of a class's implementation are quite tedious, and don't shed much light.

heh, sorry, I'm responsible for that test; it was intended to ensure that state transitions were happening as expected and the actions within the state as expected, so was sort of testing implementation details on purpose, but since it was using mocks those sorts of tests are painful when refactoring stuff. Its probably worth retaining testing of some sort to ensure the state machine behaves as it is supposed to so that new callers can't mess stuff up, but not sure the best way to do that without being fragile/implementation dependent.

paul-rogers commented:
@clintropolis, on the `SqlLifecycleTest`, one of the next PRs will refactor `SqlLifecycle` so it doesn't really have states. That one of the goals of this refactoring: once we get things in line, we can replace that class with a statement-like class that enforces its states via its API. But, that's for the next PR.

The present build again failed due to a corrupted Maven-downloaded jar file:

```text
 error reading /home/travis/.m2/repository/org/assertj/assertj-core/3.19.0/assertj-core-3.19.0.jar; zip END header not found
```

Since this has happened twice now, I wonder if the file is corrupted upstream. If so, we'll need to wait for the file to be corrected, or the cache to be flushed. Without this file working, we can't run the ITs.

clintropolis commented:
Looks like the query integration tests are legitimately failing due to changed error messages
```
The exception was thrown with the wrong message: expected ".* Parameter at position\[0] is not bound" but got "Error -1 (00000) : Remote driver error: SqlPlanningException: org.apache.druid.java.util.common.IAE: Parameter at position [0] is not bound"
	at org.apache.druid.tests.query.ITJdbcQueryTest.testJdbcPrepareStatementQueryMissingParameters(ITJdbcQueryTest.java:220)
```

paul-rogers commented:
@clintropolis, yes foiled by a single space. Fixed that and now the build is clean and ready for a final review, thanks.
